### DISCUSSION TITLE
Examples in WinAppSDK?

### DISCUSSION THREAD

**Discussion by TheJoeFin** [micjahn/ZXing.Net#540] [Q&A] [answered]
I have an app (Text Grab) which can create and read QR Codes. Currently I'm trying to make a little app which just does QR Codes, but using WinAppSDK. I'm having a hard time finding the right Writer/Renderer to write to a `Microsoft.UI.XAML.Media.BitmapImage`. I think that might be the easiest way to update the source of an Image... but I'm getting cryptic errors from WinAppSDK: `System.Runtime.InteropServices.COMException: ''`

I took the method from my WPF App and tried to do some adapting... but it is not working. This is the code generating the error:

```
public static async void SetQrCodeSourceForText(BitmapImage? bitmapImage, string text, ErrorCorrectionLevel correctionLevel)
{
    BitmapRenderer bitmapRenderer = new();
    bitmapRenderer.Foreground = Color.Black;
    bitmapRenderer.Background = Color.White;

    BarcodeWriter barcodeWriter = new()
    {
        Format = ZXing.BarcodeFormat.QR_CODE,
        Renderer = bitmapRenderer,
    };

    EncodingOptions encodingOptions = new()
    {
        Width = 500,
        Height = 500,
        Margin = 5,
    };
    encodingOptions.Hints.Add(ZXing.EncodeHintType.ERROR_CORRECTION, correctionLevel);
    barcodeWriter.Options = encodingOptions;

    Bitmap bitmap = barcodeWriter.Write(text);
    using MemoryStream ms = new();
    bitmap.Save(ms, ImageFormat.Bmp);
    bitmapImage = null;
    bitmapImage = new BitmapImage();
    await bitmapImage.SetSourceAsync(ms.AsRandomAccessStream());
}

```

**Comment by TheJoeFin** [comment, marked as answer]
Following up here. I stopped trying to do the work Async and the issue went away. Here is a complete helper which works in WinAppSDK if anyone wants to copy/paste.

```
using Microsoft.UI.Xaml.Media.Imaging;
using System.Drawing;
using System.Drawing.Imaging;
using ZXing.Common;
using ZXing.QrCode.Internal;
using ZXing.Windows.Compatibility;

public static class BarcodeHelpers
{
    public static BitmapImage GetQrCodeBitmapFromText(string text, ErrorCorrectionLevel correctionLevel)
    {
        BitmapRenderer bitmapRenderer = new();
        bitmapRenderer.Foreground = Color.Black;
        bitmapRenderer.Background = Color.White;

        BarcodeWriter barcodeWriter = new()
        {
            Format = ZXing.BarcodeFormat.QR_CODE,
            Renderer = bitmapRenderer,
        };

        EncodingOptions encodingOptions = new()
        {
            Width = 500,
            Height = 500,
            Margin = 2,
        };
        encodingOptions.Hints.Add(ZXing.EncodeHintType.ERROR_CORRECTION, correctionLevel);
        barcodeWriter.Options = encodingOptions;

        using Bitmap bitmap = barcodeWriter.Write(text);
        using MemoryStream ms = new();
        bitmap.Save(ms, ImageFormat.Png);
        BitmapImage bitmapImage = new();
        ms.Position = 0;
        bitmapImage.SetSource(ms.AsRandomAccessStream());

        return bitmapImage;
    }
}
```

**Reply by micjahn** [reply]
Thanks for sharing it.

**Comment by usedbyanr** [comment]
I realize this is an old thread but other people are going to read this. I have not tested what I am about to write but I'm pretty sure it was the cause. The original code is declared ```public static async void SetQrCodeSourceForText(BitmapImage? bitmapImage, string text, ErrorCorrectionLevel correctionLevel)``` which means this async function is going to return nothing. The call to ```await bitmapImage.SetSourceAsync(ms.AsRandomAccessStream());``` starts the async Task and returns its own Task to the caller. If the code had been declared ```public static async Task SetQrCodeSourceForText(BitmapImage? bitmapImage, string text, ErrorCorrectionLevel correctionLevel)``` and the caller used ```await SetQrCodeSourceForText(...)``` I believe it would have worked fine. Basically, the caller of an async void has no way of knowing when an async Task is complete. Async void is syntactically correct because it can be used on event handlers that are fire and forget and should only be used there.
